src/modules/upower.rs

```diff
@@ -0,0 +1,236 @@
+            .attach(None, move |properties: HashMap<String, OwnedValue>| {
+                let format = percentage_re.replace_all(&format, |_captures: &Captures| {
+                    properties["Percentage"]
+                        .downcast_ref::<f64>()
+                        .unwrap()
+                        .to_string()
+                });
+                let icon_name =
+                    String::from("icon:") + properties["IconName"].downcast_ref::<str>().unwrap();
+                label.set_markup(format.as_ref());
+                Continue(true)
+            });
+
+        let popup = self.into_popup(context.controller_tx, context.popup_rx, info);
+
+        Ok(ModuleWidget {
+            widget: container,
+            popup,
+        })
+    }
+
+    fn into_popup(
+        self,
+        _tx: mpsc::Sender<Self::ReceiveMessage>,
+        rx: glib::Receiver<Self::SendMessage>,
+        _info: &ModuleInfo,
+    ) -> Option<gtk::Box>
+    where
+        Self: Sized,
+    {
+        let container = gtk::Box::builder()
+            .orientation(Orientation::Horizontal)
+            .name("popup-upower")
+            .build();
+
+        let label = Label::builder().name("upower-details").build();
+
+        container.add(&label);
+
+        rx.attach(None, move |properties| {
+            use BatteryState::*;
+            let mut format = String::new();
+            let Some(state) = properties.get("State") else {
+                return Continue(true);
+            };
+            let state: &u32 = state.downcast_ref::<u32>().unwrap();
+            let state: BatteryState = u32_to_batterystate(*state).unwrap();
+            match state {
+                Charging | PendingCharge => {
+                    let ttf = *properties["TimeToFull"].downcast_ref::<i64>().unwrap();
+                    if ttf > 0 {
+                        format = format!("Full in {}", seconds_to_string(ttf));
+                    }
+                }
+                Discharging | PendingDischarge => {
+                    let tte = *properties["TimeToEmpty"].downcast_ref::<i64>().unwrap();
+                    if tte > 0 {
+                        format = format!("Empty in {}", seconds_to_string(tte));
+                    }
+                }
+                _ => {}
+            }
+            label.set_markup(&format);
+
+            Continue(true)
+        });
+
+        container.show_all();
+
+        Some(container)
+    }
+}
+
+fn seconds_to_string(seconds: i64) -> String {
+    let mut time_string = String::new();
+    let days = seconds / (DAY);
+    if days > 0 {
+        time_string += &format!("{days}d");
+    }
+    let hours = (seconds % DAY) / HOUR;
+    if hours > 0 {
+        time_string += &format!(" {hours}h");
+    }
+    let minutes = (seconds % HOUR) / MINUTE;
+    if minutes > 0 {
+        time_string += &format!(" {minutes}m");
+    }
+    time_string
+}
+
+fn u32_to_batterystate(number: u32) -> Result<BatteryState, u32> {
```

Comment:
This can be made constant:

```suggestion
const fn u32_to_batterystate(number: u32) -> Result<BatteryState, u32> {
```